File: Haskell/DAY5.hs
{-# LANGUAGE TypeApplications #-}
module DAY5 where

import Data.List.Split (splitOn)
import Data.List (sortOn)

main :: IO ()
main = do
    file <- readFile "../External_Files/day5.txt"
    let (rs,qs) = parseInput file
        rs' = mergeRanges rs
    print $ length (filter (\q -> any (\(b,e) -> b<=q && q<=e) rs') qs) -- PART 1: 529
    print $ sum (map (\(b,e) -> e-b+1) rs')                             -- PART 2: 344260049617193

parseInput :: String -> ([(Int,Int)], [Int])
parseInput file = (map parseRange rsStr, map (read @Int) qsStr)
    where [rsStr, qsStr] = splitOn [""] (lines file)
          parseRange s = let [a,b] = splitOn "-" s in (read a, read b)

mergeRanges :: [(Int,Int)] -> [(Int,Int)]
mergeRanges = foldl step [] . sortOn fst
  where step [] r = [r]
        step acc@((b1,e1):rest) (b2,e2)
            | e1 >= b2 - 1 = (b1, max e1 e2) : rest
            | otherwise    = (b2, e2) : acc

File: Python/DAY5.py
with open('External_Files/day5.txt') as f:
    lines = f.read().splitlines()

ranges = []
while True:
    if (line := lines.pop(0)) == '': break
    ranges.append(list(map(int, line.split('-'))))

union = []
for b,e in sorted(ranges):
    if union and union[-1][1] >= b-1: union[-1][1] = max(union[-1][1], e)
    else: union.append([b,e])

pt1 = 0
while lines:
    query = int(lines.pop(0))
    for b,e in union:
        if b<=query<=e:
            pt1 += 1
            break

pt2 = sum([e-b+1 for b,e in union])

# PART 1: 529 
# PART 2: 344260049617193
print(pt1, pt2)